worlds/witness/__init__.py

```diff
@@ -83,163 +103,87 @@ def generate_early(self):
-
-        # Remove junk, Functioning Brain, useful items (non-door), useful door items in that order until there is room
-        if itempool_difference < 0:
-            junk = [
-                item for item in pool
-                if item.classification in {ItemClassification.filler, ItemClassification.trap}
-                and item.name != "Functioning Brain"
-            ]
-
-            f_brain = [item for item in pool if item.name == "Functioning Brain"]
-
-            usefuls = [
-                item for item in pool
-                if item.classification == ItemClassification.useful
-                and item.name not in StaticWitnessLogic.ALL_DOOR_ITEMS_AS_DICT
-            ]
-
-            removable_doors = [
-                item for item in pool
-                if item.classification == ItemClassification.useful
-                and item.name in StaticWitnessLogic.ALL_DOOR_ITEMS_AS_DICT
-            ]
-
-            self.multiworld.per_slot_randoms[self.player].shuffle(junk)
-            self.multiworld.per_slot_randoms[self.player].shuffle(usefuls)
-            self.multiworld.per_slot_randoms[self.player].shuffle(removable_doors)
-
-            removed_junk = False
-            removed_usefuls = False
-            removed_doors = False
-
-            for i in range(itempool_difference, 0):
-                if junk:
-                    pool.remove(junk.pop())
-                    removed_junk = True
-                elif f_brain:
-                    pool.remove(f_brain.pop())
-                elif usefuls:
-                    pool.remove(usefuls.pop())
-                    removed_usefuls = True
-                elif removable_doors:
-                    pool.remove(removable_doors.pop())
-                    removed_doors = True
-
-            warn = make_warning_string(
-                removed_junk, removed_usefuls, removed_doors, not junk, not usefuls, not removable_doors
-            )
-
-            if warn:
-                warning(f"This Witness world has too few locations to place all its items."
-                        f" In order to make space, {warn} had to be removed.")
-
-        # Finally, add the generated pool to the overall itempool
-        self.multiworld.itempool += pool
+                # Force the item onto the tutorial gate check and remove it from our random pool.
+                self.multiworld.get_location("Tutorial Gate Open", self.player)\
+                    .place_locked_item(self.create_item(random_early_item))
+                if item_pool[random_early_item] == 1:
+                    item_pool.pop(random_early_item)
+                else:
+                    item_pool[random_early_item] -= 1
+
+        # Generate the actual items.
+        for item_name, quantity in item_pool.items():
+            self.multiworld.itempool += [self.create_item(item_name) for _ in range(0, quantity)]
+            if self.items.item_data[item_name].local_only:
```

Comment:
Note: this is not currently being used by anything live. I've tested it locally with my other changes.